### PR TITLE
Add Pyodide Test Suite

### LINKED ISSUE #238: Add tests to maintain Pyodide support
Now that Pyodide is in master, it would be good to have tests for it.

@alcarney's notes:
  * We can try using [@vscode/test-web](https://code.visualstudio.com/api/extension-guides/web-extensions#test-your-web-extension-in-a-browser-using-vscodetestweb) to test the example VSCode extension
  * Pyodide looks like it uses selenium and/or playwright to test itself, I don't see an obvious reusable component in the repo but we might be able to use their test suite as inspiration

### PR DESCRIPTION
**Pull request by alcarney**
## Description (e.g. "Related to ...", etc.)

Closes #238 

TLDR: I have the existing `pygls` test suite running but failing in a Pyodide environment and thought it's probably at a convenient place to start getting some feedback. :smile:  

image

----

After doing some more research towards #238, I came across [this issue](https://github.com/pyodide/pyodide/issues/1791) in the pyodide repo that basically says as long as your tests are packaged, you can use `pytest` from within pyodide. 

Running with this idea, this PR implements a proof of concept pyodide test suite for `pygls`, with the following workflow.

1. Build a local wheel containing `pygls` and its test suite.
1. Copy the wheel into the new `pyodide_testrunner/` folder
1. Spin up a local `http.server` pointed at the `pyodide_testrunner` folder
1. Open `http://localhost:8000` in your web browser, the tests should start running once pyodide has bootstrapped itself.

In order to get this running in CI I'm assuming we'd be able to script steps 1-3 and stand up a headless browser session somehow (selenium?) as a replacement for 4.

Now for some notes on the implementation.

---

The implementation breaks down into three main steps/commits

- Rename `tests/` -> `pygls_testsuite/`.

  This is purely to get the test suite included in the wheel when packaging, I don't particularly care about this step so happy to change it back/tweak packaging rules/include it as a sub package of `pygls` - whatever we decide the best approach to be 

- Add proof of concept test runner

  The `pyodide_testrunner/` folder contains just enough glue code to get the tests loaded and executing under pyodide.

  - It runs the tests in a web worker to keep the web page responsive while the tests run.
  - It overides `sys.stdout` in the pyodide environment so the test output can be redirected to the DOM - a nice to have but not necessary for the tests to run.
  - Invokes pytest using the [`--pyargs`](https://docs.pytest.org/en/7.1.x/how-to/usage.html#specifying-which-tests-to-run) option, which is what allows us to load tests via the `pygls_testsuite` package

- Attempt to provide `ClientServer` implementation for pyodide

  Up til now everything has worked suprisingly well! Unfortunately upon running the test suite for the first time everything errored out as the `ClientServer` object used to drive the majority of the test suite relies on threads - which are not available in pyodide!

  So this final commit is a (failing) attempt at providing an implementation of `ClientServer` compatible with the pyodide runtime.
  It works on the same principal as the glue code in the exmaple web extension from #218, but it doesn't quite work yet and I'd like to see if anyone has a better way we can handle this. - See code comments

- What doesn't work?
  
  There is a mismatch between the arguments being passed to `send_request` and what the pyodide implementation expects.
  ```
  =================================== FAILURES ===================================
  ______________________________ test_bf_initialize ______________________________

  client_server = (<Mock id='28467504'>, <pygls.server.LanguageServer object at 0x1b60698>)

      def test_bf_initialize(client_server):
          client, server = client_server
          root_uri = pathlib.Path(__file__).parent.as_uri()
          process_id = 1234
    
          response = client.lsp.send_request(
              INITIALIZE,
              {
                  "processId": process_id,
                  "rootUri": root_uri,
                  "capabilities": ClientCapabilities(),
              }
          ).result(timeout=CALL_TIMEOUT)
     
  >       assert server.process_id == process_id
  E       assert 12345 == 1234
  E        +  where 12345 = <pygls.server.LanguageServer object at 0x1b60698>.process_id

  /lib/python3.10/site-packages/pygls_testsuite/test_language_server.py:54: AssertionError
  ------------------------------ Captured log call -------------------------------
  ERROR    pygls.protocol:protocol.py:338 Failed to handle request 80d5e54f-42bf-4edd-8f3d-dfac26750d9b initialize {'processId':   1234, 'rootUri': 'file:///lib/python3.10/site-packages/pygls_testsuite', 'capabilities': ClientCapabilities(workspace=None, text_document=None, window=None, general=None, experimental=None)}
  Traceback (most recent call last):
    File "/lib/python3.10/site-packages/pygls/protocol.py", line 332, in _handle_request
      self._execute_request(msg_id, handler, params)
    File "/lib/python3.10/site-packages/pygls/protocol.py", line 261, in _execute_request
      method_name, method_type, msg_id, handler(params))
    File "/lib/python3.10/site-packages/pygls/protocol.py", line 74, in decorator
      ret_val = base_func(self, *args, **kwargs)
    File "/lib/python3.10/site-packages/pygls/protocol.py", line 601, in lsp_initialize
      self._server.process_id = params.process_id
  AttributeError: 'dict' object has no attribute 'process_id'
  ``` 
  I think this is because the pyodide implementation is bypassing the parsing machinery in `pygls` and is expected to be given objects - not sure if this is a good thing or not...

## Code review checklist (for code reviewer to complete)

- [ ] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [ ] Title summarizes what is changing
- [ ] Commit messages are meaningful (see [this][commit messages] for details)
- [ ] Tests have been included and/or updated, as appropriate
- [ ] Docstrings have been included and/or updated, as appropriate
- [ ] Standalone docs have been updated accordingly
- [ ] [CONTRIBUTORS.md][contributors] was updated, as appropriate
- [ ] Changelog has been updated, as needed (see [CHANGELOG.md][changelog])

[changelog]: https://github.com/openlawlibrary/pygls/blob/master/CHANGELOG.md
[commit messages]: https://chris.beams.io/posts/git-commit/
[contributors]: https://github.com/openlawlibrary/pygls/blob/master/CONTRIBUTORS.md

### CODE REVIEW

File: pygls_testsuite/conftest.py [hunk 1]
```diff
@@ -43,79 +48,140 @@
...
+    class ClientServer:
+
+        def __init__(self):
+
+            self.server = LanguageServer()
+            self.client = Mock()
+
+            # Pyodide specific implementations/patches.
+            self.server.lsp._send_data = self._send_data
+            self.client.lsp.send_request = self._send_request
```
Comment (alcarney) [inline pygls_testsuite/conftest.py:61]: This overrides the [usual `send_request` implementation](https://github.com/openlawlibrary/pygls/blob/master/pygls/protocol.py#L477) with [this one](https://github.com/openlawlibrary/pygls/pull/239/commits/1de4efd2139b5663a22ba1a12a52c10abd2aa0bb#diff-aee6bad608e2e8758dbcab42e829eba66f0f394079ef176817844829c7617400R85) that passes the message direct to the server's `_procedure_handler`

File: pygls_testsuite/conftest.py [hunk 2]
```diff
@@ -43,79 +48,140 @@
...
+if IS_PYODIDE:
+    class ClientServer:
+
+        def __init__(self):
+
+            self.server = LanguageServer()
+            self.client = Mock()
+
+            # Pyodide specific implementations/patches.
+            self.server.lsp._send_data = self._send_data
```
Comment (alcarney) [inline pygls_testsuite/conftest.py:60]: We also have to

**Comment by alcarney** [comment]
> In theory Pyodide should just work exactly the same as normal Python right? In practice that might not be the case, but it would be an exception

Yes, though I think the easiest way to find these exceptions is probably just to run the whole test suite within Pyodide, that way we're always checking compatibility even as new features are added. Plus I don't think it will be that difficult to do - once we find the right way to set it up (In theory it's already set up - it's just not passing yet :smile:).

> Then there would be no other modifications to the existing test framework, like changing conftest.py or creating the wheel package.

I'm not sure we can get away from creating a wheel, at the very least we would need to load `pygls` itself into Pyodide - which is usually done by installing a wheel package.

> something like Jasmine?

Definitely open to the idea, though I don't necessarily know how it compares to something like Selenium to have much of an opinion yet :smile:

**Comment by tombh** [comment]
You've got it working now!?

This approach certainly is comprehensive, and it has precedent in the Gihub Action "matrix", namely running a separate CI job for every version of Python (including Pyodide now right!). But I'm not sure I got my point across before. What about thinking about it in terms of the 80/20 rule? Therefore, what is something that is 20% of the work but gets us 80% of the results? This actually applies to the CI matrix as well: version-specific test failures happen even less than 20% of the time, but he majority of the work in surfacing bugs is done by the tests within the matrix, not the matrix itself.

So ideally 80% of the effort in this PR would just be testing the _new_ and currently _untested_ (by any test in any matrix) behaviour of the Pyodide implementation: `send_data`, threading etc. Of course if you've got the entire test suite running then no problem 😏 I'm just trying to clarify my point.

There's 2 other things that stand out for me now:
  * Duplication of `class ClientServer:`, cant that be refactored?
  * The renaming of the test folder, can that not just be renamed at build time instead? Or maybe a symlinked?

**Comment by alcarney** [comment]
> You've got it working now!?

I’m no further forward than when I opened this… I guess when I say ‘working’ I mean `pytest ….` executes as expected and produces a nice list of errors to be fixed. 

I get what you mean about the 80/20 rule…. I’m just struggling to pick on anything that’s specifically related to pyodide that would be a useful test. To me at least, the bulk of the work related to Pyodide is actually external to pygls’ implementation i.e. how to best communicate with the server? rather than anything that pygls is doing specifically. That’s mostly why I’ve focused on the existing test suite, as the `ClientServer` object is driving a server externally.

> Duplication of class `ClientServer`:, cant that be refactored?

Not sure… at least on first impressions I feel you’re going to need 2 implementions as the existing one relies on threads - which aren’t available in Pyodide. Unless of course we can come up with an implementation that would work for both… 🤔 

> The renaming of the test folder, can that not just be renamed at build time instead? Or maybe a symlinked?

I’m sure that’s possible, that first commit was just a "let me see if I can get something to work" kind of thing 😄

**Comment by alcarney** [comment]
> I’m no further forward than when I opened this…

I have been thinking about this a lot though… and I’m not convinced that [this](https://github.com/openlawlibrary/pygls/blob/96bc69df9c963979ce76f386e83fa059093bc284/pygls/protocol.py#L382) is necessarily a good idea anymore. I’m going to have a play with seeing if we can instead send things over stdio as normal (removing the need for a special case) and instead set something up in the JS ‘glue’ layer to handle the Pyodide specifics.

**Comment by alcarney** [comment]
Update: I have a (mostly) passing testsuite!

- I've [removed](https://github.com/openlawlibrary/pygls/pull/239/commits/5492e396031ad4ffd8aaac0a6223e3ff7b4d82ac) the special case for pyodide in `_send_data`, instead relying on the existing `TransportAdapter` architecture
  (**Note:** This breaks the example web extension but I'm hoping it should be easy enough to fix by adjusting the 'glue' code)
- This allowed me to write a `ClientServer` implementation that maps calls to `_send_data` on the client/server directly to `_procedure_handler` calls on the other
- I had to [disable](https://github.com/openlawlibrary/pygls/pull/239/commits/e17ec4ed881cad488326a9dab4809aae1fcb5ec7#diff-0b44323c1196f8332a32027951eadcafae3bf3641fe63b5bd8726dfc0f210663) a few [tests](https://github.com/openlawlibrary/pygls/pull/239/commits/e17ec4ed881cad488326a9dab4809aae1fcb5ec7#diff-6d5da8fba067bfcec7caeddc8da06c20815d444b4c38b3c095e98d7739c745f0) as they seem to involve threads in some capacity, but the majority seems to just work! :smile: 
- Still need to figure out some of the workflow stuff, like undoing the test suite rename etc - but hopefully this is a step in the right direction? 

```
============================= test session starts ==============================
platform emscripten -- Python 3.10.2, pytest-7.1.1, pluggy-1.0.0
rootdir: /home/pyodide
plugins: asyncio-0.18.3, typeguard-2.13.3
asyncio: mode=legacy
collecting ... 
collecting 0 items                                                             
collecting 51 items                                                            
collecting 106 items                                                           
collecting 162 items                                                           
collected 167 items                                                            

test_document.py ...............                                         [  8%]
test_feature_manager.py ...........                                      [ 15%]
test_language_server.py ..sss..                                          [ 19%]
test_protocol.py .............                                           [ 27%]
test_server_connection.py ss                                             [ 28%]
test_types.py ...                                                        [ 30%]
test_uris.py .......sssss                                                [ 37%]
test_workspace.py .......                                                [ 41%]
lsp/test_call_hierarchy.py .....                                         [ 44%]
lsp/test_code_action.py ...                                              [ 46%]
lsp/test_code_lens.py ...                                                [ 48%]
lsp/test_color_presentation.py ..                                        [ 49%]
lsp/test_completion.py ..                                                [ 50%]
lsp/test_declaration.py .....                                            [ 53%]
lsp/test_definition.py .....                                             [ 56%]
lsp/test_document_color.py ..                                            [ 58%]
lsp/test_document_highlight.py ...                                       [ 59%]
lsp/test_document_link.py ...                                            [ 61%]
lsp/test_document_symbol.py ...                                          [ 63%]
lsp/test_folding_range.py ...                                            [ 65%]
lsp/test_formatting.py ...                                               [ 67%]
lsp/test_hover.py ....                                                   [ 69%]
lsp/test_implementation.py .....                                         [ 72%]
lsp/test_linked_editing_range.py ...                                     [ 74%]
lsp/test_moniker.py ...                                                  [ 76%]
lsp/test_on_type_formatting.py ...                                       [ 77%]
lsp/test_prepare_rename.py ....                                          [ 80%]
lsp/test_progress.py ..                                                  [ 81%]
lsp/test_range_formatting.py ...                                         [ 83%]
lsp/test_references.py ...                                               [ 85%]
lsp/test_rename.py ...                                                   [ 86%]
lsp/test_selection_range.py ...                                          [ 88%]
lsp/test_semantic_tokens.py ...........                                  [ 95%]
lsp/test_signature_help.py ...                                           [ 97%]
lsp/test_type_definition.py .....                                        [100%]

=============================== warnings summary ===============================
../../lib/python3.10/site-packages/pytest_asyncio/plugin.py:191
  /lib/python3.10/site-packages/pytest_asyncio/plugin.py:191: DeprecationWarning: The 'asyncio_mode' default value will change to 'strict' in future, please explicitly use 'asyncio_mode=strict' or 'asyncio_mode=auto' in pytest configuration file.
    config.issue_config_time_warning(LEGACY_MODE, stacklevel=2)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================== 157 passed, 10 skipped, 1 warning in 4.24s ==================
```

**Comment by alcarney** [comment]
[Screencast from 06-07-22 23:58:41.webm](https://user-images.githubusercontent.com/2675694/177657005-0b424242-39cb-43db-b3de-5ea3a95f0c9b.webm)

(Hopefully the above screencast works for you, it seems to freeze in the middle for me...)

I think it's getting there! There is now a python script that automates the process of building a wheel containing pygls + testsuite, spinning up a local http server and using selenium to run the tests in a browser. Currently it's hard coded to use Chrome/Chromium.

Stuff still to finish
- Update CI to include pyodide, tweak test script to launch headless browsers in CI
- Fix example extension
- Docs? I'm sure we'd all appreciate some notes on all this, let me know if you want anything in particular included.

Let me know if you have any feedback :smile:

**Comment by tombh** [comment]
Looking good! This sort of thing, with building, packaging, browsers, and more, is always hard to juggle. So it's great that you've got something working already.

Which docs are you thinking about? [The existing ones for the fountain extension](https://github.com/openlawlibrary/pygls/tree/master/examples/fountain-extension)?

**Comment by alcarney** [comment]
>Which docs are you thinking about? [The existing ones for the fountain extension](https://github.com/openlawlibrary/pygls/tree/master/examples/fountain-extension)?

More about documenting how the test suite and `pyodide_test_runner` are setup.

**Comment by alcarney** [comment]
I've moved both `ClientServer` implementations into `tests/ls_setup.py` :)

**Comment by tombh** [comment]
Really nice :) I'm happy with this PR now then. So we just need to resolve those conflicts cos I merged that big Flakey Tests Fix PR. Looking at them I reckon I could give the resolutions a go if you want? Or if they seem straightforward to you then no problem if you want to give it a try too

**Comment by alcarney** [comment]
I think I've resolved everything correctly, maybe you might want to take a quick look and double check things?

**Review by tombh — APPROVED** [review]

**Comment by tombh** [comment]
Awesome! 🎉🤸
